### ISSUE TITLE
Incorrect error messages of type parameters

### ISSUE THREAD

**Issue by falsandtru** [Microsoft/TypeScript#6248]
```ts
interface Foo<T> {
    n: T;
}
interface Foo<U> {
    y: U;
}

interface FooConstructor {
	new <T>(): Foo<T>;
}
```

```
TS2314 Generic type 'Foo<T, U>' requires 2 type argument(s).
```

Feedback of https://github.com/DefinitelyTyped/DefinitelyTyped/pull/7321. I searched `Promise<T, R>` interface but I couldn't found it because it does not exist anywhere. Other contributors will may have a same difficulty in supporting of compatibility with es6(lib.es6.d.ts) definitions.

**Comment by DanielRosenwasser** [comment]
You should have gotten another error "All declarations of an interface must have identical type parameters." The second error message is caused by the first. Just rename `U` to `T` and that should fix both errors.

**Comment by DanielRosenwasser** [comment]
Actually, reopening - if we can improve things so that the cascading error doesn't give worse messages down the line, we should try to do so.

**Comment by falsandtru** [comment]
Thank you for your accepting. Actually, I couldn't found an another error because I got too many errors larger than 2000. It will be difficult that we find your specified error during refactoring of DefinitelyTyped provided declaration files.

**Comment by DavidKDeutsch** [comment]
Ah, so glad I found this; I've been banging my head trying to figure out why I was getting the "All declarations of 'Promise' must have identical type parameters" error. Since all of the cascading errors referred to 'Promise<T, R>' I was convinced that there was an definition somewhere that took two type arguments, not that the problem was that Promise<T> conflicts with Promise<R>.

Out of curiosity, why is this an error? AFAIK the name of a generic argument is only used locally by the implementing code, i.e. code that uses it only needs to know about the position/constraints.

**Comment by RyanCavanaugh** [comment]
Confirmed fix, we have an error on the differing declarations now. AI said

## Unable to Reproduce ❌

I attempted to reproduce this issue but could not confirm it still exists.

Test case used:
**interfaces.ts:**
```typescript
interface Foo<T> {
    n: T;
}
interface Foo<U> {
    y: U;
}

interface FooConstructor {
    new <T>(): Foo<T>;
}

```

**tsconfig.json:**
```json
{
    "compilerOptions": {
        "strict": true
    }
}
```


**Command:** `tsc -p .`
**TypeScript Version:** Version 5.9.0-dev.20250722
**Exit Code:** 2

<details>
<summary>Compiler Output</summary>

**Stdout:**
```
C:/Users/ryanca/AppData/Local/Temp/repro-check-EbsJXe/interfaces.ts(1,11): error TS2428: All declarations of 'Foo' must have identical type parameters.
C:/Users/ryanca/AppData/Local/Temp/repro-check-EbsJXe/interfaces.ts(4,11): error TS2428: All declarations of 'Foo' must have identical type parameters.
C:/Users/ryanca/AppData/Local/Temp/repro-check-EbsJXe/interfaces.ts(9,16): error TS2314: Generic type 'Foo<T, U>' requires 2 type argument(s).

```

**Stderr:**
```
(empty)
```
</details>

**What you observed in the bug report:** The error message stated 'Generic type 'Foo<T, U>' requires 2 type argument(s).' This was confusing because it suggested needing two type arguments instead of addressing the root issue of inconsistent generic type parameters in the merged interfaces.
**What you expected:** The expected behavior is an error informing that all declarations of an interface must have identical type parameters when merging interfaces with generics.
**What I observed when running tsc:** The error messages are 'All declarations of 'Foo' must have identical type parameters' and 'Generic type 'Foo<T, U>' requires 2 type argument(s).'. The first error correctly identifies the actual issue.

My reasoning:
> The original bug report indicated that the user was facing a confusing error message. Specifically, the user saw a message requiring two type arguments for a merged interface, instead of getting a clear message that the interface declarations should have identical type parameters.

The expected behavior is for the compiler to provide an informative error message that correctly identifies that the issue is with having different type parameters in the merged interfaces, rather than providing a confusing message that hints at non-existent type arguments.

In the current compiler output, we see `TS2428` errors indicating that "All declarations of 'Foo' must have identical type parameters," which is exactly the error message the user was hoping to see initially. Additionally, the `TS2314` error still appears, but it follows the helpful `TS2428` errors instead of being the sole or first error message.

The helpful error messages explaining that the declarations of 'Foo' must have identical type parameters are indeed what is expected, and they are now given in the current compiler output. Therefore, the confusing aspect of the error message (the one faced by the user) has been resolved.


---

**Analysis Result:**
```json
{
  "bug_is_present": "no",
  "files": [
    {
      "filename": "interfaces.ts",
      "content": "interface Foo<T> {\n    n: T;\n}\ninterface Foo<U> {\n    y: U;\n}\n\ninterface FooConstructor {\n    new <T>(): Foo<T>;\n}\n"
    },
    {
      "filename": "tsconfig.json",
      "content": "{\n    \"compilerOptions\": {\n        \"strict\": true\n    }\n}"
    }
  ],
  "user_reported_behavior": "The error message stated 'Generic type 'Foo<T, U>' requires 2 type argument(s).' This was confusing because it suggested needing two type arguments instead of addressing the root issue of inconsistent generic type parameters in the merged interfaces.",
  "expected_behavior": "The expected behavior is an error informing that all declarations of an interface must have identical type parameters when merging interfaces with generics.",
  "current_observed_behavior": "The error messages are 'All declarations of 'Foo' must have identical type parameters' and 'Generic type 'Foo<T, U>' requires 2 type argument(s).'. The first error correctly identifies the actual issue.",
  "command_line": "tsc -p .",
  "tsc_version": "Version 5.9.0-dev.20250722",
  "compiler_output": {
    "stdout": "C:/Users/ryanca/AppData/Local/Temp/repro-check-EbsJXe/interfaces.ts(1,11): error TS2428: All declarations of 'Foo' must have identical type parameters.\r\nC:/Users/ryanca/AppData/Local/Temp/repro-check-EbsJXe/interfaces.ts(4,11): error TS2428: All declarations of 'Foo' must have identical type parameters.\r\nC:/Users/ryanca/AppData/Local/Temp/repro-check-EbsJXe/interfaces.ts(9,16): error TS2314: Generic type 'Foo<T, U>' requires 2 type argument(s).\r\n",
    "stderr": "",
    "exit_code": 2
  },
  "reasoning": "The original bug report indicated that the user was facing a confusing error message. Specifically, the user saw a message requiring two type arguments for a merged interface, instead of getting a clear message that the interface declarations should have identical type parameters.\n\nThe expected behavior is for the compiler to provide an informative error message that correctly identifies that the issue is with having different type parameters in the merged interfaces, rather than providing a confusing message that hints at non-existent type arguments.\n\nIn the current compiler output, we see `TS2428` errors indicating that \"All declarations of 'Foo' must have identical type parameters,\" which is exactly the error message the user was hoping to see initially. Additionally, the `TS2314` error still appears, but it follows the helpful `TS2428` errors instead of being the sole or first error message.\n\nThe helpful error messages explaining that the declarations of 'Foo' must have identical type parameters are indeed what is expected, and they are now given in the current compiler output. Therefore, the confusing aspect of the error message (the one faced by the user) has been resolved."
}
```
